### PR TITLE
feat(staker): add CancelExternalIncentive for pre-start incentives

### PR DESCRIPTION
**Pull request by jinoosss**
## Summary

Adds `CancelExternalIncentive`, an admin/governance-only entry point on the staker that terminates an external incentive **before it starts**, removing the incentive and refunding the reward tokens and the GNS deposit to a refund address.

`EndExternalIncentive` already exists, but it only runs once the incentive's end timestamp has passed. An incentive created with a wrong pool, reward token, or amount therefore stayed funded until its entire window elapsed. This PR fills that gap for the pre-start window.

## Why removal (not the `refunded` flag) is safe before start

`EndExternalIncentive` keeps the record and sets a `refunded` flag because deposits and the penalty accounting still reference a started incentive. Before the start timestamp, nothing references the incentive:

- Both `StakeToken` and the lazy reward discovery scan the per-pool start-time index only up to the current time, so an incentive with `start > now` is never attached to any deposit.
- Once removed from the incentive tree, the per-pool `byStartTime` index, and the global incentive tree, every stake/collect/getter path behaves as if it had never been created.

So the cancel path removes the record outright, restricted to `now < startTimestamp`. A defensive guard additionally rejects any incentive that already carries a distributed reward or accumulated penalty.

## Impact on related functions

| Function | Behavior after cancel |
|---|---|
| `StakeToken` | A pending incentive is never discovered (scan upper bound is `now`); index removal prevents the "id in index but missing from tree" panic. If it was the pool's only incentive, the pool becomes non-stakeable again (expected). |
| `calculatePositionReward` / `CollectReward` | Pending incentives never enter a deposit's id set, so removal is invisible to reward calculation. |
| `UnStakeToken` / `CollectReward` | Do not depend on incentive existence; positions staked earlier can still unstake and collect. |
| `CreateExternalIncentive` | Incentive id counter is monotonic, so a replacement created after a cancel gets a fresh id. |

## Changes

- **feat(staker)**: `CancelExternalIncentive` in the impl, proxy, and `IStaker` interface; `assertIsNotStartedIncentive`; removal primitives (`RemoveIncentive`, `RemoveIncentiveByStartTime`, resolver `remove`); error `GNOSWAP-STAKER-023`. Reward refund is capped by the staker balance with a shortfall event (mirrors `EndExternalIncentive`), and removal happens before transfers to preserve CEI.
- **feat(gov)**: register the staker `CancelExternalIncentive` handler in the governance parameter-change registry — otherwise the governance half of the `AssertIsAdminOrGovernance` check was unreachable.
- **test(staker)**: unit tests (guard rails + success path + same-start-time sibling preservation + balance-capped refund + accrued-incentive rejection) and 4 scenario filetests (start-timestamp boundary, multiple incentives on one pool, permission + Safe/Emergency halt, cancel-and-recreate).
- **test(gov)**: scenario filetest running the cancel through a full propose/vote/execute governance flow.
- **docs**: both staker READMEs and `docs/staker.md`.
- Upgrade fixtures (`v2_valid`, `v3_valid`, `v2_invalid`, `mock`, `security`) updated to satisfy the extended `IStaker` interface.

## Testing

Individually passing: `staker`, `staker/v1`, `gov/governance`, `gov/governance/v1`, and scenario packages `staker`, `gov/governance`, `upgrade`, `halt`, `getter`.

> Note: running the whole `.../...` tree in one `gno test` invocation fails `staker/v1` with `version_manager: initializer already registered` — a pre-existing init-time registry collision unrelated to this change (all packages pass when run individually, and CI runs package-by-package).

## Summary by CodeRabbit

- **New Features**
  - Authorized administrators or governance can cancel external incentives strictly before they start.
  - Cancelled incentives are removed from active listings, with available rewards and the GNS deposit refunded to the incentive creator.
  - Refunds are capped by available balances, with shortfalls reported.
  - External incentives can be created by callers who provide the required tokens.
  - Added validation for token availability, pools, authorization, timing, incentive status, accrued rewards, and penalties.
- **Documentation**
  - Clarified cancellation requirements, refund handling, and incentive lifecycle rules.
- **Tests**
  - Added coverage for authorization, validation, refunds, indexing, boundary conditions, and related edge cases.

### CODE REVIEW

File: contract/r/gnoswap/staker/v1/external_incentive.gno [hunk 1]
```diff
@@ -287,6 +287,119 @@ func (s *stakerV1) endExternalIncentive(resolver *PoolResolver, incentiveResolve
...
+	incentivesResolver.remove(incentiveResolver.ExternalIncentive)
+	s.getExternalIncentives().remove(incentiveId)
+
+	if refundAmount > 0 {
+		common.SafeGRC20Transfer(cross(rlm), rewardToken, refundAddress, refundAmount)
+	}
+
+	if depositGnsAmount > 0 {
+		gns.Transfer(cross(rlm), refundAddress, depositGnsAmount)
+	}
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply contract/r/gnoswap/staker/v1/external_incentive.gno:392]: 31e07202e0e44e62e4b25b7ad4c74424d72470f3

**Review by junghoon-vans — APPROVED** [review]
